### PR TITLE
New module: keycloak_group

### PR DESCRIPTION
**Pull request by adamgoossens**
##### SUMMARY
This module (keycloak_group) allows using the Keycloak REST API (http://www.keycloak.org/docs-api/3.4/rest-api/index.html) to administrate Keycloak (http://www.keycloak.org) groups. It allows users to create, update and remove groups.

Keycloak is an Open Source Identity and Access Management system spearheaded by Red Hat. It provides OpenID Connect and SAML authentication/authorization services.

Two prior modules in this namespace have been merged (keycloak_client in #31716 and keycloak_clienttemplate in #33419). This module reuses and extends code introduced therein (from module_utils/keycloak.py).

##### ISSUE TYPE
New Module Pull Request

##### COMPONENT NAME
keycloak_group

##### ANSIBLE VERSION
2.8.0

### CODE REVIEW

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: This seems to be a leftover, there is no state "updated" anywhere else in the code.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: When required is false, it can be left out (since it is the default).

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py:77]: Are there any attributes commonly available and with defined semantics in, for instance, Keycloak 3.4, or is this completely freeform user-definable?

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py:81]: This should be fine; if and when you want to implement this, please have a look at the check_role_representation in #35558, it might be helpful :)

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: Does this register do anything, i.e. is it used somewhere?

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: I will steal this.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: This branch can never be hit since the argument_spec defines which states are valid and checks the provided module arguments against it.

File: lib/ansible/module_utils/keycloak.py
Comment (eikef) [inline lib/ansible/module_utils/keycloak.py]: probably lowercase gid here :)

**Review by eikef — CHANGES_REQUESTED** [review]
Code looks good, with some minor changes.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: after_group does not exist at this point, this will stacktrace

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (eikef) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: This is missing a state, so the group is created instead of being deleted.

**Comment by eikef** [comment]
The module does what it is supposed to do -- creation, deletion, and updates of groups all work. There are some minor code issues, once they are addressed this looks good to me to ship.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py:77]: Totally user-definable key value pairs for storing additional custom data on the group.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: Ah, of course - thanks!

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: ....yep, it certainly is. Will fix :)

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: I've used it in test playbooks to retrieve the auto-generated group ID. I'll remove it since it doesn't really value add to the example.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: Good point - just me being overcautious. Will remove.

**Comment by adamgoossens** [comment]
ready_for_review

**Review by eikef — APPROVED** [review]

**Comment by eikef** [comment]
shipit

**Comment by getjack** [comment]
This PR seems to be stale. Can we still shipit?

**Comment by RickS-C137** [comment]
[shipit]

**Comment by adamgoossens** [comment]
The code itself should still be good, but I will need to amend the version tag - currently marked as available from Ansible 2.5 :)

**Comment by resmo** [comment]
I would like to merge this into 2.8. 

However, I don't like the the "return". It returns 3 states of the same thing: wanted, existing, end_stateTo me this looks over complicated. I would suggest to only return the "real values" and document them properly.

This way, integration tests can easily be done as we do in other modules:

- create in check mode --> expect empty dict return and changed=true
- create in live mode --> return dict with values and changed=true
- create in live mode for idempotency test --> return dict with values and changed=false

See https://github.com/ansible/ansible/blob/devel/test/integration/targets/vultr_user/tasks/main.yml as an example. Any thoughts?

**Comment by resmo** [comment]
needs_info

**Comment by adamgoossens** [comment]
Hi @resmo , I have no problems with that. I will spend some time on this this weekend and get it ready for review again in time for 2.8.

**Comment by adamgoossens** [comment]
ready_for_review

**Comment by resmo** [comment]
well done. I only see a few minor cosmetic things in docs which should not block this from merging, I'll comment inline.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (resmo) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: space looks unnecessary

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (resmo) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: shouldn't this be  group instead of group_id?

~~~diff
 -    result = dict(changed=False, msg='', diff={}, group_id='')
 +    result = dict(changed=False, msg='', diff={}, group='')
~~~

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (resmo) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: this should be type complex and document every possible return in this, see https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/hcloud/hcloud_ssh_key.py#L88 as an example

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (resmo) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: `msg` is a common return and not module specific and must not be documented. Please remove

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (resmo) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py]: not a blocker but `local_action` is a thing the ansible_linter suggest to change. That is why I would prefer:

~~~yaml

- name: Update the name of a Keycloak group
  keycloak_group:
    id: '9d59aa76-2755-48c6-b1af-beb70a82c3cd'
    name: an-updated-kc-group-name
    state: present
    auth_client_id: admin-cli
    auth_keycloak_url: https://auth.example.com/auth
    auth_realm: master
    auth_username: USERNAME
    auth_password: PASSWORD
  delegate_to: localhost
~~~

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (resmo) [inline lib/ansible/modules/identity/keycloak/keycloak_group.py:48]: not a blocker but we started to add a `type: (bool|str|dict|int)` for each option as well, this would look like 

~~~diff
         default: 'present'
+        type: str
~~~

**Review by resmo — CHANGES_REQUESTED** [review]
Please let me know if you shouldn't have time to work on these minors before the freeze on 4.4.19

**Comment by adamgoossens** [comment]
> Please let me know if you shouldn't have time to work on these minors before the freeze on 4.4.19

Thanks for the comprehensive review! I'll tackle these tomorrow my time. Should be no problem to have them completed by end of this weekend, ahead of the freeze.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py:48]: Thanks, I've added types for each of the module options.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: Done. Thanks!

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: This is now done. The Keycloak API isn't particularly well documented, however this particular set of return values has been pretty stable over the last few versions. I don't expect description for this complex return value to change.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: Agreed - fixed.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: Agreed - removed.

File: lib/ansible/modules/identity/keycloak/keycloak_group.py
Comment (adamgoossens) [reply lib/ansible/modules/identity/keycloak/keycloak_group.py]: Agreed - I've now changed and pushed this. Previous commit missed the delegate_to option.

**Comment by adamgoossens** [comment]
ready_for_review

**Review by resmo — APPROVED** [review]
LGTM

**Comment by adamgoossens** [comment]
Thanks @resmo - hasn't seem to have merged however; is there anything else I need to do before 04 Apr? @eikef does it also need another tick from you as the maintainer?

**Comment by resmo** [comment]
We have a little more time because the freeze has been delayed to 11.04. I would be good if @eikef give a final 'shipit'. However, new modules won't get automerged and I have to merge it. 

I realized that the existing modules in for keycloak also have this "proposed, existing, end_state", I would suggest not to continue this existing return pattern and follow the way you did in this module.

I will merge it, if we don't get a "veto" by @eikef

**Comment by resmo** [comment]
rebuild_merge

**Comment by dazasa** [comment]
Hi, do you intend to implement the role mapping in this script? Best regards David

**Pull request merged**
